nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java

```diff
@@ -141,48 +191,64 @@ public void setProperty(final String name, final String value) {
 
-        lock.lock();
-        try {
-            verifyModifiable();
-
-            try (final NarCloseable narCloseable = NarCloseable.withComponentNarLoader(component.getClass())) {
-                final PropertyDescriptor descriptor = component.getPropertyDescriptor(name);
-                String value = null;
-                if (!descriptor.isRequired() && (value = properties.remove(descriptor)) != null) {
-
-                    if (descriptor.getControllerServiceDefinition() != null) {
-                        if (value != null) {
-                            final ControllerServiceNode oldNode = serviceProvider.getControllerServiceNode(value);
-                            if (oldNode != null) {
-                                oldNode.removeReference(this);
-                            }
-                        }
-                    }
+        final PropertyDescriptor descriptor = component.getPropertyDescriptor(name);
+        String value = null;
+        if (!descriptor.isRequired() && (value = properties.remove(descriptor)) != null) {
 
-                    try {
-                        component.onPropertyModified(descriptor, value, null);
-                    } catch (final Exception e) {
-                        // nothing really to do here...
+            if (descriptor.getControllerServiceDefinition() != null) {
+                if (value != null) {
+                    final ControllerServiceNode oldNode = serviceProvider.getControllerServiceNode(value);
+                    if (oldNode != null) {
+                        oldNode.removeReference(this);
                     }
-
-                    return true;
                 }
             }
-        } finally {
-            lock.unlock();
+
+            try {
+                component.onPropertyModified(descriptor, value, null);
+            } catch (final Exception e) {
+                // nothing really to do here...
+            }
+
+            return true;
         }
+
         return false;
     }
 
+    /**
+     * Adds all of the modules identified by the given module paths to the InstanceClassLoader for this component.
+     *
+     * @param modulePaths a list of module paths where each entry can be a comma-separated list of multiple module paths
+     */
+    private void processClasspathModifiers(final Set<String> modulePaths) {
+        try {
+            final URL[] urls = ClassLoaderUtils.getURLsForClasspath(modulePaths, null, true);
+
+            final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+            if (!(classLoader instanceof InstanceClassLoader)) {
+                // Really shouldn't happen, but if we somehow got here and don't have an InstanceClassLoader then log a warning and move on
+                logger.warn("Unable to modify the classpath for {}, expected InstanceClassLoader, but found {}",
+                        new Object[] { name, classLoader.getClass().getName()});
+                return;
```

Comment:
Agree with the first scenario.
As for the ```ClassLoaderUtils.getURLsForClasspath(...)```, I've read what you describe and I personally lean to the side of fail fast. Don't get me wrong; what you describe is perfectly valid, but I approach it very differently where the _fatal_ condition I am referring to will most definitely be a flow developer's error which will be caught during the development of the flow and therefore is not expected to happen in prod, so delegating to the component developer to do custom validation in each and every new/existing component seems like on overhead.